helix-term/src/ui/menu.rs

```diff
@@ -92,36 +95,70 @@ impl<T: Item> Menu<T> {
             recalculate: true,
         };
 
-        // TODO: scoring on empty input should just use a fastpath
-        menu.score("");
-
+        menu.re_score(true);
         menu
     }
 
-    pub fn score(&mut self, pattern: &str) {
+    pub fn re_score(&mut self, reset_cursor: bool) {
+        let prev_matched_option = if !reset_cursor {
+            self.cursor
+                .and_then(|c| self.matches.get(c).map(|(index, _)| *index))
+        } else {
+            None
+        };
+
         // reuse the matches allocation
         self.matches.clear();
-        self.matches.extend(
-            self.options
-                .iter()
-                .enumerate()
-                .filter_map(|(index, option)| {
-                    let text: String = option.filter_text(&self.editor_data).into();
-                    // TODO: using fuzzy_indices could give us the char idx for match highlighting
-                    self.matcher
-                        .fuzzy_match(&text, pattern)
-                        .map(|score| (index, score))
-                }),
-        );
-        // matches.sort_unstable_by_key(|(_, score)| -score);
-        self.matches.sort_unstable_by_key(|(index, _score)| {
-            self.options[*index].sort_text(&self.editor_data)
-        });
+        // shortcut, if there isn't a pattern
+        if self.previous_pattern.is_empty() {
+            self.matches
+                .extend(self.options.iter().enumerate().map(|(o, _)| (o, 0)))
+        } else {
+            self.matches.extend(
+                self.options
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(index, option)| {
+                        let text: String = option.filter_text(&self.editor_data).into();
+                        // TODO: using fuzzy_indices could give us the char idx for match highlighting
+                        self.matcher
+                            .fuzzy_match(&text, &self.previous_pattern)
+                            .map(|score| (index, score))
+                    }),
+            );
+            // matches.sort_unstable_by_key(|(_, score)| -score);
+            self.matches.sort_unstable_by_key(|(index, _score)| {
+                self.options[*index].sort_text(&self.editor_data)
+            });
```

Comment:
Is it a good idea to run `sort_text` within a sort?

Reply:
What do you mean exactly?
This is actually untouched code. Just for clarification `sort_text` doesn't actually sort the text, but returns the text to be used for sorting